### ISSUE TITLE
Implement action conditionals that check for items in inventory

### ISSUE THREAD

**Issue by Dumplingz** [uchicago-cs/chiventure#577]
Continues implementation of issue #505, of the modules described [here](https://github.com/uchicago-cs/chiventure/wiki/Action-Management-Inventory-Item-Prerequisites). Conditionals are now split into multiple types: `attribute_condition` and `inventory_condition`. The logic for inventory_condition now needs to be implemented and specifically player and game-state must be modified and relevant function names will be changed.

**Comment by Dumplingz** [comment]
I'm trying to include the "player" struct in item.h, since all the condition code is in there. However, since player.h includes item.h, and item.h now must include player.h, the compiler cannot figure out which to do first and cannot compile.

**Comment by Dumplingz** [comment]
> I'm trying to include the "player" struct in item.h, since all the condition code is in there. However, since player.h includes item.h, and item.h now must include player.h, the compiler cannot figure out which to do first and cannot compile.

I'm thinking a separate condition.h file is in order

**Comment by Dumplingz** [comment]
Just got this error after running ctest on Linux:
6: criterion: Could not initialize the message server: Address already in use.
6/6 Test #6: test-wdl .........................Child aborted***Exception:   0.03 sec
It passed all other tests, just this one. What's weird is that I'm unable to reproduce it, since another 4-5 runs of ctest --extra-verbose all gave me the all clear. Hopefully it's not a big deal

**Comment by Dumplingz** [comment]
> Just got this error after running ctest on Linux:
> `6: criterion: Could not initialize the message server: Address already in use. 6/6 Test #6: test-wdl .........................Child aborted***Exception: 0.03 sec`
> It passed all other tests, just this one. What's weird is that I'm unable to reproduce it, since another 4-5 runs of ctest --extra-verbose all gave me the all clear. Hopefully it's not a big deal

This error seems to be able to happen to all tests, but it looks like it may be Criterion itself
`1: criterion: Could not initialize the message server: Address already in use.
1/6 Test #1: test-action_management ...........Child aborted***Exception:   0.01 sec
test 2`

**Comment by Dumplingz** [comment]
Had the error happen on tests 1,4,5,6

**Comment by Dumplingz** [comment]
> I'm trying to include the "player" struct in item.h, since all the condition code is in there. However, since player.h includes item.h, and item.h now must include player.h, the compiler cannot figure out which to do first and cannot compile.

I ended up putting the typedef of `player_t` into `item.h`. It works, but there must be a better solution.

**Comment by Dumplingz** [comment]
I'm really confused about the add_action_condition function. It has 5 arguments:
`item_t *item`
`game_action_t *action`
`item_t *cond_item`
`attribute_t *cond_attribute`
`attribute_value_t cond_value`

However, `cond_item` is only checked if it's null. Also, `item` and `action` are used only in the function `get_action(item, action->action_name);` in order to get the action. I'm having a hard time figuring out why we can't just use `action` directly, rather than finding the action from `item` and `action->action_name`.

I propose to revise this function by removing `get_action` and removing `item` since they seem superfluous.

**Comment by Dumplingz** [comment]
As I expected, removing a few of the function arguments still allows it to compile in the same way.

**Comment by Dumplingz** [comment]
> Had the error happen on tests 1,4,5,6

And 3... not sure what's going on

**Comment by Dumplingz** [comment]
> > Had the error happen on tests 1,4,5,6
> 
> And 3... not sure what's going on. Testing on ssh'd UC linux

After running ctest many times, I've found that the error can happen in all tests (1-6)

**Comment by Dumplingz** [comment]
> > > Had the error happen on tests 1,4,5,6
> > 
> > 
> > And 3... not sure what's going on. Testing on ssh'd UC linux
> 
> After running ctest many times, I've found that the error can happen in all tests (1-6)

Checked out the dev branch and the error can appear also.

**Comment by Dumplingz** [comment]
Getting another problem with freeing... it's tied to player_free

**Comment by Dumplingz** [comment]
delete_all_items is causing it, and I'm not sure how to fix it.

**Comment by Dumplingz** [comment]
Much work on abstraction needs to be done before this can be correctly merged into dev.
To Do:
1)
all_conditions_met currently only takes an `item_t *item` and `char *action_name`, whereas I would like it to take an `action_condition_list_t` and return SUCCESS or FAILURE. Requires a lot of refactoring.

2)
There are two (or more) different and mutually exclusive ways of implementation which I must choose one of. Both functions described below are in game.c 
1. (What I will likely support, since I think it provides more abstraction) Currently, the **end condition** checker `add_end_condition_to_game` is written to take a `game_t` and `game_action_condition_t`.
2. The old add_condition (referring to a condition prerequisite for an **action**) that has a `game_t` struct, `char * action_name`, and all of the input variables for creating a new attribute_condition.
In the future, if new types of conditions are required (more do exist, such as a condition that checks if a room has been visited or not #238), it will be much nicer to only have one add_condition function that supports all variations of conditions.

3)
All condition functions/structs should not be in item.h; it doesn't make sense, especially with end conditions that have nothing to do with item.h.

**Comment by Dumplingz** [comment]
Since conditionals are so complicated and general, I have added a new condition.c and condition.h files and ported all the relevant functions/structs there.

**Comment by Dumplingz** [comment]
I've attempted to implement a totally new way of doing conditions, similar to the way the newly added end conditions are currently implemented. Rather than have specific functions that take in a "item, attribute_name, attribute_value," I instead split it up into the more general attribute_condition_new, which accommodates those inputs, ore the new inventory_condition_new which accommodates inventory conditionals. However, both output the  only a condition_t struct, so thus all subsequent functions will be able to take only a condition_t struct and there will be no repeated code or unnecessary extra functions for different types of conditions.

**Comment by Dumplingz** [comment]
Many, many include problems. Somehow item.h isn't compiled, and I get this message for everything:
`error: unknown type name 'item_hash_t'`

**Comment by Dumplingz** [comment]
> Many, many include problems. Somehow item.h isn't compiled, and I get this message for everything:
> `error: unknown type name 'item_hash_t'`

I was able to fix this issue by adding a couple forward declarations.

**Comment by Dumplingz** [comment]
> > Many, many include problems. Somehow item.h isn't compiled, and I get this message for everything:
> > `error: unknown type name 'item_hash_t'`
> 
> I was able to fix this issue by adding a couple forward declarations.

Nevermind, still more include issues. I'm seeing a lot of problems similar to the original player_t struct definition.

**Comment by Dumplingz** [comment]
Everything compiles (finally)! On to writing more tests for the the new condition files

**Comment by Dumplingz** [comment]
PR #772 has been merged, so this issue is being closed.

This issue has successfully added a new kind of condition that checks if a given item is in a player's inventory or not. However, in the process of implementing this, a rework of the condition structures was also done in order to accommodate this new kind of condition and facilitate the implementation of other kinds of conditions.

First, the new files condition.c and condition.h were created, which now house all of the logic for conditions, moving it from game_action.c/h and item.c/h.

The new struct `condition` was created, which contains either an `attribute_condition` or an `inventory_condition` pointer, with a union tag to identify which condition it is for the logic. This struct is created by either `attribute_condition_new` or `inventory_condition_new`, which take in the necessary inputs for each condition.

Many functions, including `add_end_condition_to_game`, `all_conditions_left` and `add_condition` were generalized to take a `condition` struct (generated from the two functions mentioned in the last point) so that future conditions added don't have to change anything about these functions.

Since all condition logic should be in the condition files now, theoretically #238 can be almost completely implemented only in condition.c/h.

Tests were added in test_item_item_actions.c (in action management's tests) and the new file test_condition.c (in game state's tests).

This issue also uncovered a bug in freeing a `player` struct, described in detail in #785. WDL's part of issue #236 is described in #786.

**Comment by MaxineK36** [comment]
Issue Score: ✔️++

Comments:
Great work! The closing statement on this issue is particularly well done.
